src/snowflake/snowpark/_internal/debug_utils.py

```diff
@@ -198,3 +199,55 @@ def get_df_transform_trace_message(
             "environment variable."
         )
     return "\n".join(debug_info_lines)
+
+
+def get_python_source_from_sql_error(error_msg: str, args: tuple) -> Optional[str]:
+    """
+    Extract SQL error line number and map it back to Python source code. We use the
+    helper function get_plan_from_line_numbers to get the plan from the line number
+    found in the SQL compilation error message. We then extract the source lines
+    and columns using the ast_id associated with the plan.
+    """
+    sql_compilation_error_regex = re.compile(
+        r""".*SQL compilation error:\s*error line (\d+) at position (\d+).*""",
+    )
+    match = sql_compilation_error_regex.match(error_msg)
+    if not match:
+        return None
+
+    sql_line_number = int(match.group(1)) - 1
+    from snowflake.snowpark._internal.analyzer.snowflake_plan import SnowflakePlan
+    from snowflake.snowpark._internal.analyzer.select_statement import (
+        Selectable,
+    )
+
+    for arg in args:
+        if isinstance(arg, SnowflakePlan) or isinstance(arg, Selectable):
+            try:
+                from snowflake.snowpark._internal.utils import (
+                    get_plan_from_line_numbers,
+                )
+
+                plan = get_plan_from_line_numbers(arg, sql_line_number)
+                ast_id = None
+                bind_stmt = None
+                if isinstance(plan, Selectable) and plan.df_ast_ids is not None:
+                    ast_id = plan.df_ast_ids[-1]
+                    bind_stmt = plan._session._ast_batch._bind_stmt_cache.get(ast_id)
```

Comment:
I think in this case, when there are multiple ast_ids associated, we should list sources of all dataframes as possible source of error